### DISCUSSION TITLE
WASD/Arrows Support for Teleop Panel

### DISCUSSION THREAD

**Discussion by mcdoerr** [foxglove/community#221] [Q&A]
Our current project uses the WASD keys to drive the robot instead of clicking buttons with the mouse. We are looking to switch to Foxglove because it fits our use case well, but don't want to lose this important functionality. 

We would like to request support for driving the robot with Foxglove's teleop panel using either WASD keys or the arrow keys. The keys could map to the existing inputs for the four buttons in the screenshot when the screen is in focus. 

This would greatly reduce the difficulty of driving our robot, since it is much easier to click keys than to move the mouse, especially on faster robots.

**Comment by mcdoerr** [comment]
![sample_teleop](https://user-images.githubusercontent.com/45296521/151439958-2a856ed4-f13f-46e8-8a63-47dc4bd99d2b.png)
For example, the WASD keys could map like this. The arrow keys would work just as well.

**Comment by jtbandes** [comment]
Thanks for the feature request. One issue we have often run into with panel keyboard shortcuts is the issue of supporting multiple panels. If a user configures a layout with 2 teleop panels (possibly for multiple robots) which one should the keyboard shortcuts apply to? Do you have any thoughts on this from the perspective of your project?

**Comment by mcdoerr** [comment]
That issue makes sense, thanks for bringing it up!

In our current simulation tool, we deal with multiple teleop panels by having the user select the panel which they are driving in with a mouse click. When they click off of the "driving panel" the keyboard inputs do not send messages until the panel is reselected/focused. If a user had multiple teleop panels, they could switch by clicking on the different panels. This also does not affect the logic for driving with the mouse clicks, which is nice.

Perhaps it would be best to have this feature turned off by default, and an option to turn it on in settings? What are your thoughts? 

Thanks for the quick response! :)

**Reply by Tobias-Pe** [reply]
Backing @mcdoerr up:

I also like this idea. So only the currently focused (shown by the violet border color) one should be controllable by keyboard.

![image](https://github.com/foxglove/community/assets/49952224/6e0bc277-3c09-4e98-b819-8dfcf198a07a)

@jtbandes what do you think ?

**Comment by robi772** [comment]
Please ADD a STOP button into the middle.

**Comment by jtbandes** [comment]
Hi @robi772, can you clarify what you would want the stop button to do? Currently, the Teleop panel only publishes messages while the buttons are pressed, so just releasing the button stops the flow of messages. Would the stop button be sending a different message?

**Comment by robi772** [comment]
I like to control a robot with the turtlebot3_teleop_key, when continously publishing the linear and angular messages. You just set the speed with the buttons.

**Reply by Tobias-Pe** [reply]
Your idea is changing more how to think about the teleop panel.

This discussion here just makes the teleop panel controllable with the keyboard without changing the logic inside of it.

I still can see the usefulness of a stop button and the logic u proposed, but i would see it in another disccusion-thread.

**Comment by Tobias-Pe** [comment]
@jtbandes 
The teleop panel would be so much more usefull if you could use your keyboard's up, down, left, right arrows.
Therefore, you could also press two buttons at the same time, instead of using the gui buttons.

In the documentation under migration, it says the teleop panel should be usable instead of this tool: http://wiki.ros.org/teleop_twist_keyboard
Currently, the teleop_twist_keyboard is way more versatile and practical.

I would also love that, and it would make the teleop panel way more usefull ^^
